### PR TITLE
Added fixer for PHPUnit's @expectedException annotation

### PR DESCRIPTION
**Pull request by ro0NL**
See https://github.com/symfony/symfony/pull/20043 and https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/2028

### CODE REVIEW

**Comment by SpacePossum** [comment]
LGTM :)

Not sure about the name as it is kind of vague IMHO (i.e. does this fixer remove multiple `@` like `@@Expec..`, does it fix the casing of the tag like `@ExpectedException`? I know it doesn't and shouldn't ;)) and all the classnames of the PHPDoc fixers are prefixed with `Phpdoc` ATM.

**Comment by ro0NL** [comment]
What about `PhpUnit\PhpdocExpectedExceptionToFqcnFixer`

**Comment by SpacePossum** [comment]
nice one :)
I expect the tests to fail because you need to re-gen the readme though ;)

**Comment by phansys** [comment]
This should run before `no_unused_imports`, WDYT?

**Comment by ro0NL** [comment]
About the readme.. im not sure but the diff is pretty huge on my machine when running
`php-cs-fixer readme > README.rst`

```diff
 * **align_double_arrow**
-                         Align double arrow symbols in
-                         consecutive lines.
+                           Align double arrow symbols
+                           in consecutive lines.

etc.
```

@phansys yes..

**Comment by SpacePossum** [comment]
the big diff is because the name you picked for the fixer is longer than the current longest so the indent changes :)
no worries, re-gen and commit ^-^

**Comment by phansys** [comment]
The diff is fine for sure @ro0NL, since the command re generates the whole file (re aligning all fixer's descriptions by instance) according the changes you made. You can test it ensuring it's fine in your local environment by running `phpunit`.

**Comment by ro0NL** [comment]
Should we call this `PhpUnitPhpdocExpectedExceptionWithRootNamespaceFixer`? :sweat_smile:

File: src/Fixer/PhpUnit/PhpdocExpectedExceptionWithRootNamespaceFixer.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+namespace PhpCsFixer\Fixer\PhpUnit;
+
+use PhpCsFixer\AbstractFixer;
+use PhpCsFixer\Tokenizer\Tokens;
+
+/**
+ * @author Roland Franssen <franssen.roland@gmail.com>
+ */
+class PhpdocExpectedExceptionWithRootNamespaceFixer extends AbstractFixer
```
Comment (keradus) [inline src/Fixer/PhpUnit/PhpdocExpectedExceptionWithRootNamespaceFixer.php]: final

File: src/Fixer/PhpUnit/PhpdocExpectedExceptionWithRootNamespaceFixer.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+     * {@inheritdoc}
+     */
+    public function fix(\SplFileInfo $file, Tokens $tokens)
+    {
+        foreach ($tokens as $token) {
+            if (!$token->isGivenKind(T_DOC_COMMENT)) {
+                continue;
+            }
+
+            $token->setContent(preg_replace_callback('~^(\s*\*\s*@expectedException\s+)([^\s\\\\].*)$~m', function (array $matches) {
```
Comment (keradus) [inline src/Fixer/PhpUnit/PhpdocExpectedExceptionWithRootNamespaceFixer.php]: why callback version ?
why set content when regexp won't change anything ?

**Comment by keradus** [comment]
> This should run before `no_unused_imports`, WDYT?

then the integration test is missing

File: src/Fixer/PhpUnit/PhpdocExpectedExceptionWithRootNamespaceFixer.php (hunk 2 above)
Comment (ro0NL) [inline src/Fixer/PhpUnit/PhpdocExpectedExceptionWithRootNamespaceFixer.php]: This is simple string normalization.. what do you propose?

Comment (ro0NL) [inline src/Fixer/PhpUnit/PhpdocExpectedExceptionWithRootNamespaceFixer.php]: Im not fully aware of inner workings, but i noticed calling `setContent` is marking tokens as changed. so yes, lets avoid that when content is not changed :+1:

**Comment by ro0NL** [comment]
The readme breaks at https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2200/files#diff-88b99bb28683bd5b7e3a204826ead112R197 

see https://github.com/ro0NL/PHP-CS-Fixer/blob/4cfee8b13364bc130ef80c32bd508b79641b37a1/README.rst

File: src/Fixer/PhpUnit/PhpUnitExpectedExceptionAnnotationWithRootNamespaceFixer.php [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    {
+        foreach ($tokens as $token) {
+            if (!$token->isGivenKind(T_DOC_COMMENT)) {
+                continue;
+            }
+
+            $content = $token->getContent();
+            $newContent = preg_replace('~^(\s*\*\s*@expectedException\s+)([^\s\\\\].*)$~m', '$1\\\\$2', $content);
+
+            if ($newContent !== $content) {
```
Comment (SpacePossum) [inline src/Fixer/PhpUnit/PhpUnitExpectedExceptionAnnotationWithRootNamespaceFixer.php]: check is not needed,
https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/master/src/Tokenizer/Token.php#L479

File: tests/Fixtures/Integration/priority/php_unit_expected_exception_annotation_with_root_namespace,no_unused_imports.test [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+use Some\Exception;
+
+class Foo extends \PHPUnit_Framework_TestCase
+{
+    /**
+     * @expectedException Exception
+     */
+    public function testBar()
+    { }
+}
```
Comment (SpacePossum) [inline tests/Fixtures/Integration/priority/php_unit_expected_exception_annotation_with_root_namespace,no_unused_imports.test]: please add the prio test to this list/test as well;
https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/master/tests/FixerFactoryTest.php#L193

File: src/Fixer/PhpUnit/PhpdocExpectedExceptionWithRootNamespaceFixer.php [hunk 5]
```diff
@@ -31,6 +31,15 @@ public function getDescription()
     /**
      * {@inheritdoc}
      */
+    public function getPriority()
+    {
+        // should be run before NoUnusedImportsFixer
+        return -9;
```
Comment (keradus) [inline src/Fixer/PhpUnit/PhpdocExpectedExceptionWithRootNamespaceFixer.php]: add test here as well: https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/master/tests/FixerFactoryTest.php#L278

**Comment by ro0NL** [comment]
Ready? :) Do we need a 1.12 backport for symfony?

File: README.rst [hunk 6]
```diff
@@ -175,657 +175,1079 @@ automatically fix anything:
...
+                                        MUST be
+                                        enclosed by
+                                        braces. Braces
+                                        should be
+                                        properly
+                                        placed. Body
+                                        of braces
+                                        should be
+                                        properly
+                                        indented.
```
Comment (keradus) [inline README.rst]: i dont like putting one word per statement...
sth to be resolved in new PR.

nothing to worry about for you @ro0NL ;)

**Comment by keradus** [comment]
add to Symfony ruleset (https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/master/src/RuleSet.php#L54), also update Symfony integration test to cover this new fixer as well

**Comment by keradus** [comment]
https://github.com/ro0NL/PHP-CS-Fixer/pull/1

**Comment by phansys** [comment]
Why not to use the naming defined at the [PHP's standard resolution rules](http://php.net/manual/en/language.namespaces.rules.php) instead?

**Comment by ro0NL** [comment]
@phansys because it doesnt apply to `@expectedException`, it's intended to be a FQCN. However i agree this is often unclear and causes false positives like https://github.com/symfony/symfony/pull/20028/files#diff-63ff7c7532d7434f326bdd6dfd265f35

The missing root namespace is an editor problem (PHPStorm doesnt recognize it as a known class), the assertion itself works as phpunit doesnt care (it assumes FGCN anyway).

Normalizing to FQCN with root namespace shows the wrong use statements (due `no_unused_imports`) which should make things clear enough..

**Comment by keradus** [comment]
Please rebase.

Do you want to prepare version for 1.x line ?

**Comment by keradus** [comment]
yes, ubot runs 1.x

> I initially targeted master because of "new feature"

Why you didn't read this: https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/1.12/CONTRIBUTING.md#quick-guide ?

If you care about having the fixer released soon you should target 1.x, if not then 2.x is OKish

**Comment by ro0NL** [comment]
Aargh, my bad :) ill try to rebase 1.12 later.. until then 2.x is ok.

**Comment by phansys** [comment]
@ro0NL, sorry, maybe I've not explained my point in a clear way regarding the name. What about to call it `PhpUnit\PhpdocExpectedExceptionFqcnFixer`? I'm against the concept of "root" here, since it isn't used at as part of the namespace name definitions.

**Comment by ro0NL** [comment]
@phansys agree. Should we still include "annotation" in the name.. so it's clear this is not about `setExpectedException()`.

On 2nd thought, _should_ this be about `setExpectedException()`? We could also fix/normalize calls like `setExpectedException('Some\Exception');` to `'\Some\Exception'` for the sake of consistency.

**Comment by keradus** [comment]
@ro0NL Please put whole work in one line (1.x ?) and freeze the second one (2.x ?)

**Comment by ro0NL** [comment]
Closing in favor of #2208.

**Pull request closed without merging**